crypto/rand/rand_vms.c

```diff
@@ -28,106 +36,389 @@
-    for (pitem = item; pitems_data->length != 0; pitem++) {
-        pitem->length = pitems_data->length;
-        pitem->code   = pitems_data->code;
-        pitem->buffer = &data_buffer[total_length];
-        pitem->retlen = 0;
-        /* total_length is in longwords */
-        total_length += pitems_data->length / 4;
-        pitems_data++;
-    }
-    pitem->length = pitem->code = 0;
+    /* Take all the 64-bit items first, to avoid possible alignment problems */
+    total_length += prepare_item_list(JPI_item_data_64bit,
+                                      OSSL_NELEM(JPI_item_data_64bit),
+                                      JPI_items, &data_buffer[total_length]);
+    total_length += prepare_item_list(RMI_item_data_64bit,
+                                      OSSL_NELEM(RMI_item_data_64bit),
+                                      RMI_items, &data_buffer[total_length]);
+    /* Now the 32-bit items */
+    total_length += prepare_item_list(DVI_item_data, OSSL_NELEM(DVI_item_data),
+                                      DVI_items, &data_buffer[total_length]);
+    total_length += prepare_item_list(JPI_item_data, OSSL_NELEM(JPI_item_data),
+                                      JPI_items, &data_buffer[total_length]);
+    total_length += prepare_item_list(RMI_item_data, OSSL_NELEM(RMI_item_data),
+                                      RMI_items, &data_buffer[total_length]);
+    total_length += prepare_item_list(SYI_item_data, OSSL_NELEM(SYI_item_data),
+                                      SYI_items, &data_buffer[total_length]);
 
     /* Fill data_buffer with various info bits from this process */
-    if (sys$getjpiw(EFN$C_ENF, NULL, NULL, item, &iosb, 0, 0) != SS$_NORMAL)
-        return 0;
+    {
+        uint32_t status;
+        uint32_t efn;
+        IOSB iosb;
+        $DESCRIPTOR(SYSDEVICE,"SYS$SYSDEVICE:");
 
-    /* Now twist that data to seed the SSL random number init */
-    for (i = 0; i < total_length; i++) {
-        sys$gettim((struct _generic_64 *)&sys_time[0]);
-        srand(sys_time[0] * data_buffer[0] * data_buffer[1] + i);
-
-        if (i == (total_length - 1)) { /* for JPI$_FINALEXC */
-            ptr = &data_buffer[i];
-            for (j = 0; j < 4; j++) {
-                data_buffer[i + j] = ptr[j];
-                /* OK to use rand() just to scramble the seed */
-                data_buffer[i + j] ^= (sys_time[0] ^ rand());
-                tmp_length++;
-            }
-        } else {
-            /* OK to use rand() just to scramble the seed */
-            data_buffer[i] ^= (sys_time[0] ^ rand());
+        if ((status = sys$getdviw(EFN$C_ENF, 0, &SYSDEVICE, DVI_items,
+                                  0, 0, 0, 0, 0)) != SS$_NORMAL) {
+            lib$signal(status);
+            return 0;
+        }
+        if ((status = sys$getjpiw(EFN$C_ENF, 0, 0, JPI_items, 0, 0, 0))
+            != SS$_NORMAL) {
+            lib$signal(status);
+            return 0;
+        }
```

Comment:
But doesn't split means that prepare_item_list would *grow* JPI_items, not overwrite 64-bit queries with 32-bit ones? Or you'd need pair of getjpiw calls...

Reply:
Yup, I realised that just now.  Fix coming up